### PR TITLE
M3 1974 state update

### PR DESCRIPTION
**Pull request by rjstires**
I genuinely apologize for the size and reach of this pull request. For what it's worth, I think it represents an significant improvement to the data layer and user experience.

## Purpose
The purpose of this PR is to frontload users Linodes into the application state. On load of the application a Thunk is fired off to request all Linodes and store them and an array of their IDs into the application state. The array of IDs is for easy `IDs.includes(myLinodeId)` searching.

Additionally I added events middleware that dispatches additional events given a particular event. In this case the adding, deleting, or varios update linode action types. I believe this to be the future of what drives this application. Middleware can now handle the management of state, as it was intended. This makes the consuming components easier to reason about as they have no further need of events$.

Once Linodes were in store I then updated Linodes landing. First was merging in-progress events to the currently viewed Linodes. Since state contains both linodes and events, it was a matter of mapping over the currently viewed linodes and adding the in progress event found in the state.

Then it was off to updating pagination. This was incredibly simple. I created an HOC to provide the handlers, which now just created an ordered "view" of the state. I did extract data and pagination to it's own LinodeLanding specific HOC and we may be able to further extract it.

So pagination still stuttered very, very hard. It was the types requests. Every linode row was getting its own type from the API. This was an existing bug, but was not somehow more apparent. I opted to tackle moving all types (including legacy) into the store, render the typesContext moot. Which is how I ended up removing typesContext across the project.

## Funny Story
There was a random `/` on App.tsx.
image

### CODE REVIEW

**Comment by Jskobos** [comment]
Should backups drawer be fully functional at this point? Drawer is empty when it opens, and on Dashboard the CTA says "You currently have 7 Linodes without backups" which is greater than the # of Linodes on my account.

File: src/features/Dashboard/Dashboard.tsx [hunk 1]
```diff
@@ -76,15 +76,14 @@ export class Dashboard extends React.Component<CombinedProps, {}> {
 
 const mapStateToProps: MapStateToProps<StateProps, {}, ApplicationState> = (state, ownProps) => ({
   accountBackups: pathOr(false, ['__resources', 'accountSettings', 'data', 'backups_enabled'], state),
-  linodesWithoutBackups: pathOr([],['backups', 'data'], state),
+  linodesWithoutBackups: state.__resources.linodes.entities.filter(l => !l.backups.last_backup),
```
Comment (Jskobos) [inline src/features/Dashboard/Dashboard.tsx]: This is different logic than in BackupsCTA. Should probably go with that one (l.backups.enabled)

File: src/features/linodes/LinodesDetail/LinodeResize/LinodeResize.tsx [hunk 2]
```diff
@@ -200,13 +196,20 @@ export class LinodeResize extends React.Component<CombinedProps, State> {
...
-  deprecatedTypesData: (data || []).map(LinodeResize.extendType).filter((eachType) => {
-    return eachType.successor !== null;
-  })
+interface WithTypesProps {
+  currentTypesData: ExtendedType[];
+  deprecatedTypesData: ExtendedType[];
+}
+
+const withTypes = connect((state: ApplicationState, ownProps) => ({
+  currentTypesData: state.__resources.types.entities
```
Comment (Jskobos) [inline src/features/linodes/LinodesDetail/LinodeResize/LinodeResize.tsx]: Might be able to combine these two filters with R.partition

File: src/features/linodes/LinodesLanding/LinodesLanding.test.tsx [hunk 3]
```diff
@@ -58,18 +55,23 @@ describe('ListLinodes', () => {
...
             setDocs={setDocs}
             clearDocs={clearDocs}
-            typesRequest={jest.fn}
-            typesLoading={false}
-            typesLastUpdated={1}
             actions={actions}
             linodesWithoutBackups={[]}
             managed={false}
+            handleOrderChange={jest.fn()}
+            handlePageChange={jest.fn()}
```
Comment (Jskobos) [inline src/features/linodes/LinodesLanding/LinodesLanding.test.tsx]: Just a refactor suggestion, you can declare a `props` object with these fields, then spread them into the `shallow` call, updating as needed for each test case.

File: src/features/linodes/LinodesDetail/LinodeResize/LinodeResize.tsx (hunk 2 above)
Comment (rjstires) [reply src/features/linodes/LinodesDetail/LinodeResize/LinodeResize.tsx]: This is existing and I really didnt want to touch it quite yet.

File: src/features/linodes/LinodesLanding/LinodesLanding.test.tsx (hunk 3 above)
Comment (rjstires) [reply src/features/linodes/LinodesLanding/LinodesLanding.test.tsx]: Indeed

**Comment by martinmckenna** [comment]
screen shot 2018-12-11 at 3 05 04 pm

screen shot 2018-12-11 at 3 06 50 pm


Some linodes don't have a power status, even on the detail page

I'm also getting many `Warning: Encountered two children with the same key, `11183590`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.` errors upon creating a Linode

**Comment by martinmckenna** [comment]
I tested this by Creating a Linode from a Linode and then tried to reboot a different Linode

Now, upon rebooting a Linode, the calls to GET /events are going back and forth.....honestly, I'm not really sure where to begin troubleshooting. Not sure if this is an issue with the API or our X-Filters

First call to get events

screen shot 2018-12-11 at 3 18 46 pm

Second call:

screen shot 2018-12-11 at 3 19 41 pm

File: src/features/linodes/LinodesLanding/withPaginatedLinodes.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  return {
+    linodesRequestLoading: loading,
+    linodesRequestError: error,
+    linodesCount: entities.length,
+    linodesData: x,
+    linodesWithoutBackups: entities.filter(l => !l.backups.enabled),
+  }
+};
+
+const orderList: <T>(order: Order, orderBy: keyof T) => (list: T[]) => T[] = (order, orderBy) => (list) => {
```
[the remark this answers is not in the document]
Comment (rjstires) [reply src/features/linodes/LinodesLanding/withPaginatedLinodes.ts]: Yes. @johnwcallahan is looking into reselect for his work and Ill adopt it here.

File: src/features/linodes/LinodesDetail/LinodeResize/LinodeResize.tsx [hunk 5]
```diff
@@ -200,13 +196,20 @@ export class LinodeResize extends React.Component<CombinedProps, State> {
...
+    .filter((eachType) => eachType.successor === null)
+    .map(LinodeResize.extendType),
+
+  depcrecatedTypesDAta: state.__resources.types.entities
```
Comment (johnwcallahan) [inline src/features/linodes/LinodesDetail/LinodeResize/LinodeResize.tsx]: Spelling/Capitalization here, esp. because this prop is used elsewhere.

**Comment by martinmckenna** [comment]
Reposting here for visibility so it doesn't get lost later:

"And also...this is super nitpicky, but I think it's weird UX that we respond to an event on `linode_delete` because you delete your linode -> come back to the landing page -> see your Linode for a second -> then it goes away"

And the condition for when the following logic runs should be updated, so we don't do unnecessary sorting

```js
  const x = compose(
    enhanceLinodes(notifications, events),
    createDiplayPage(page, pageSize),
    orderList<Linode.Linode>(order, orderBy),
  )(entities);
```

**Pull request merged**